src/microbe_stage/MicrobeAI.cs

```diff
@@ -65,7 +70,8 @@ public void Think(float delta, Random random, MicrobeAICommonData data)
 
         ChooseActions(random, data);
 
-        // Clear the absorbed compounds for run and rumble
+        // Store & clear the absorbed compounds for run and rumble
+        previouslyAbsorbedCompounds = new Dictionary<Compound, float>(microbe.TotalAbsorbedCompounds);
```

Comment:
Since this is always set as a clone of TotalAbsorbedCompounds, is this nessicary? I see no place where this dictionary is modified.

Reply:
It is, actually, because TotalAbsorbedCompound is reset right afterwards (l.75).
This clone is here to keep the information in memory for one decision loop, because it will be used for next iteration to computation.

Reply:
Hmm, ok. I vaguely recall that; probably a future refactor candidate...

Reply:
Would it be possible to modify the behaviour so that a copy of the data is not needed? The AI can decide when it wants to clear the absorbed compounds data. If this is kept as is, maybe this should be modified to not allocate new dictionaries all the time, instead clearing previouslyAbsorbedCompounds and adding the data from TotalAbsorbedCompounds into that existing object.

Reply:
I don't really get this point : "The AI can decide when it wants to clear the absorbed compounds data."
If the AI is to compute a gradient, they need to have access to value from two ticks. I think a buffer structure is required then (at least I do not see how to circumvene this).

Removing reallocation seems definitely possible however, I'll dive onto it. Is there an interest other than memory usage?

Reply:
See my comment I posted on this PR itself, but basically, as the absorbed compounds is cleared each AI update, it *already* contains the delta relative to the previous update (as only the newly absorbed compound amounts are stored in that dictionary).
So the information on how much the AI absorbed during the last AI update cycle should already be computed in that variable and be ready to use.